sdks/python/apache_beam/runners/portability/spark_runner_test.py

```diff
@@ -0,0 +1,148 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from __future__ import absolute_import
+from __future__ import print_function
+
+import argparse
+import logging
+import sys
+import unittest
+from shutil import rmtree
+from tempfile import mkdtemp
+
+from apache_beam.options.pipeline_options import DebugOptions
+from apache_beam.options.pipeline_options import PortableOptions
+from apache_beam.runners.portability import portable_runner
+from apache_beam.runners.portability import portable_runner_test
+
+if __name__ == '__main__':
+  # Run as
+  #
+  # python -m apache_beam.runners.portability.spark_runner_test \
+  #     --spark_job_server_jar=/path/to/job_server.jar \
+  #     [SparkRunnerTest.test_method, ...]
+
+  parser = argparse.ArgumentParser(add_help=True)
+  parser.add_argument('--spark_job_server_jar',
+                      help='Job server jar to submit jobs.')
+  parser.add_argument('--environment_type', default='docker',
+                      help='Environment type. docker or process')
+  parser.add_argument('--environment_config', help='Environment config.')
+  parser.add_argument('--extra_experiments', default=[], action='append',
+                      help='Beam experiments config.')
+  known_args, args = parser.parse_known_args(sys.argv)
+  sys.argv = args
+
+  spark_job_server_jar = known_args.spark_job_server_jar
+  environment_type = known_args.environment_type.lower()
+  environment_config = (
+      known_args.environment_config if known_args.environment_config else None)
+  extra_experiments = known_args.extra_experiments
+
+  # This is defined here to only be run when we invoke this file explicitly.
+  class SparkRunnerTest(portable_runner_test.PortableRunnerTest):
+    _use_grpc = True
+    _use_subprocesses = True
+
+    @classmethod
+    def _subprocess_command(cls, job_port, expansion_port):
+      # will be cleaned up at the end of this method, and recreated and used by
+      # the job server
+      tmp_dir = mkdtemp(prefix='sparktest')
+
+      try:
+        return [
+            'java',
+            '-Dbeam.spark.test.reuseSparkContext=true',
+            '-jar', spark_job_server_jar,
+            '--spark-master-url', 'local',
+            '--artifacts-dir', tmp_dir,
+            '--job-port', str(job_port),
+            '--artifact-port', '0',
+            '--expansion-port', str(expansion_port),
+        ]
+      finally:
+        rmtree(tmp_dir)
+
+    @classmethod
+    def get_runner(cls):
+      return portable_runner.PortableRunner()
+
+    def create_options(self):
+      options = super(SparkRunnerTest, self).create_options()
+      options.view_as(DebugOptions).experiments = [
+          'beam_fn_api'] + extra_experiments
+      options._all_options['parallelism'] = 1
+      options._all_options['shutdown_sources_on_final_watermark'] = True
```

Comment:
You probably don't need those for Spark.

Reply:
The Spark runner doesn't even support them -- looks like I got a little overzealous with the copying :laughing: